### PR TITLE
feat: highlight table of content on scroll

### LINKED ISSUE #3235: Dynamically update table of contents section highlighting based on active scroll position
**Describe the bug**

I was studying the ERC-20 tutorial and noticed that when I scrolled down to **THE SELL FUNCTION** from **THE BUY FUNCTION** header, the scroll was not changing dynamically. 

**To Reproduce**

1. Go to [the dev tutorial](https://ethereum.org/en/developers/tutorials/transfers-and-approval-of-erc-20-tokens-from-a-solidity-smart-contract/#the-buy-function)

**Expected behavior**
It should be scrolling up/down dynamically depending on the position of the user inside the view port height and width. It will help in tutorials involving a lot of headers as subtopics, so that the user knows which topic in the tutorial he/she is currently on. and navigate to another subtopic when needed.

**Screenshots**

image

**Desktop (please complete the following information):**

- OS: Linux 
- BrowserFirefox
- Version 89.0

### PR DESCRIPTION
**Pull request by il3ven**
Highlight the link in table of content that is currently in viewport.

## Description
https://user-images.githubusercontent.com/4337699/126549847-ab397df6-41aa-4ff1-98c3-1c14682bb7ab.mp4

Created an useActiveHash hook. It changes the state to the id of the element that scrolled into the viewport.

## Related Issue

Resolves #3235

### CODE REVIEW

**Review by wackerow — APPROVED** [review]
@il3ven This looks great! ❤️ Very nice upgrade to the ToC on the docs, thanks so much for doing this

@corwintines or @pettinarip This  LGTM, but mind being a second set of eyes on the logic here? 🙏🏻

File: src/hooks/useActiveHash.js [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  useEffect(() => {
+    const observer = new IntersectionObserver(
+      (entries) => {
+        entries.forEach((entry) => {
+          if (entry.isIntersecting) {
+            setActiveHash(`#${entry.target.id}`)
+          }
+        })
+      },
+      { rootMargin: rootMargin || `0% 0% -80% 0%` }
```
Comment (pettinarip) [inline src/hooks/useActiveHash.js:21]: Why the `-80%` is needed?

Comment (pettinarip) [reply src/hooks/useActiveHash.js:21]: Lets set the `0% 0% -80% 0%` value as the default arg value.

```js
export const useActiveHash = (itemIds, rootMargin = `0% 0% -80% 0%`) => {
  ...
}
```

**Review by pettinarip — COMMENTED** [review]
Thanks for the PR @il3ven! I just added one small comment and one question to understand it better. Overall it looks great 💯

File: src/hooks/useActiveHash.js (hunk 1 above)
Comment (il3ven) [reply src/hooks/useActiveHash.js]: Using rootMargin we can grow or shrink the root element. The callback is fired when the observing element interacts with the root element (browser viewport in this case). 

By setting the bottom margin to -80% we ensure that the callback is fired when the heading is in the top 20% of the browser viewport. You can observe that the highlight in table of content doesn't change when the heading is visible. It changes when the heading is scrolled to the top of the viewport.

Comment (il3ven) [reply src/hooks/useActiveHash.js]: > Lets set the 0% 0% -80% 0% value as the default arg value.

Done 👍🏽

Comment (pettinarip) [reply src/hooks/useActiveHash.js]: Oh I see, thanks for the explanation!

**Review by pettinarip — APPROVED** [review]

**Review by wackerow — APPROVED** [review]
@il3ven 👍🏻 
Thanks again for posting this!

**Comment by wackerow** [comment]
@all-contributors please add @il3ven for code

**Comment by samajammin** [comment]
Looks awesome! Thanks again for the contribution @il3ven!

**Pull request merged**